state/protocol/protocol_state/kvstore_storage.go

```diff
@@ -0,0 +1,62 @@
+package protocol_state
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage/badger/transaction"
+)
+
+// ProtocolKVStore persists different snapshots of the Protocol State's Key-Calue stores [KV-stores].
+// Here, we augment the low-level primitives provided by `storage.ProtocolKVStore` with logic for
+// encoding and decoding the state snapshots into abstract representation `protocol_state.KVStoreAPI`.
+//
+// At the abstraction level here, we can only handle protocol state snapshots, whose data models are
+// supported by the current software version. There might be serialized snapshots with legacy versions
+// in the database, that are not supported anymore by this software version and can only be retrieved
+// as versioned binary blobs via `storage.ProtocolKVStore`.
```

Comment:
> Do we actually want to support this [retrieving snapshots that we can't decode with the current software]?

That is a good question. I am not sure. Some thoughts (and maybe an explanation why the code looks the way it does):
* The ability to retrieve version-encoded snapshots from `storage.ProtocolKVStore`  that the current implementation might not be able to deserialize (provided such snapshots are in the data base) was already the case. It is nothing new added by this PR. 
* However, with this PR, the documentation now explicitly mentions this, which I think sparked your comment Yurii (?) Because, explicitly mentioning it feels more like this is a properly supported feature. 
   * In my mind, the documentation reflects the de-facto abilities of the code. The separation of the `storage.ProtocolKVStore` and `protocol_state.ProtocolKVStore` arose purely from needing to modularize the implementation. 
   * given that we have `storage.ProtocolKVStore` and `protocol_state.ProtocolKVStore`, I think it is important to clarify the difference between both. 

In https://github.com/onflow/flow-go/pull/5644/commits/8e1ed9928ba28dde99fd67dfffba839b78febbdf I have updated the API documentation slightly to clarify support of legacy snapshots:

* for `protocol_state.ProtocolKVStore`, I removed the statement that legacy snapshots can be retrieved via the lower-level `storage.ProtocolKVStore` 
* in `storage.ProtocolKVStore`'s doc, I have emphasized that retrieving legacy snapshots is currently permitted, but this might change in the future.